packages/portfolio-planner/src/engine.ts

```diff
@@ -41,20 +48,222 @@ const encodedKeyToPath = (key: string) => {
 };
 
+/**
+ * Determine whether a dot-separated path starts with a sequence of path
+ * components.
+ */
+const vstoragePathStartsWith = (path: string, prefix: string) =>
+  path === prefix ||
+  path.startsWith(prefix.endsWith('.') ? prefix : `${prefix}.`);
+
 const stripPrefix = (prefix: string, str: string) => {
-  str.startsWith(prefix) || Fail`${str} is missing prefix ${prefix}`;
+  str.startsWith(prefix) || Fail`${str} is missing prefix ${q(prefix)}`;
   return str.slice(prefix.length);
 };
 
+/**
+ * Parse input as JSON, or handle an error (for e.g. substituting a default or
+ * applying a more specific message).
+ */
 const tryJsonParse = (json: string, replaceErr?: (err?: Error) => unknown) => {
   try {
     return JSON.parse(json);
   } catch (err) {
     if (!replaceErr) throw err;
-    return replaceErr(err);
+    try {
+      return replaceErr(err);
+    } catch (newErr) {
+      if (!newErr.cause) assert.note(newErr, err.message);
+      throw newErr;
+    }
   }
 };
 
+/**
+ * Map the elements of an array to new values, skipping elements for which the
+ * mapping results in either `undefined` or `false`.
+ */
+const partialMap = <T, U>(
+  arr: T[],
+  mapOrDrop: (value: T, index?: number, arr?: T[]) => U | undefined | false,
+): U[] =>
+  arr.flatMap((el, i, arrArg) => {
+    const result = mapOrDrop(el, i, arrArg);
+    if (result === undefined || result === false) return [];
+    return [result];
+  });
+
+/**
+ * Consume an async iterable with at most `capacity` unsettled results at any
+ * given time, passing each input through `processInput` and providing
+ * [result, index] pairs in settlement order.
+ * Source order can be recovered with consumer code like:
+ * ```
+ * const results = [];
+ * for (const [result, i] of makeWorkPool(...)) results[i] = result;
+ * ```
+ * or something more sophisticated to eagerly detect complete subsequences
+ * immediately following a previous high-water mark for contiguous results.
+ *
+ * To support cases in which `processInput` is used only for side effects rather
+ * than its return value, the returned AsyncGenerator has a promise-valued
+ * `done` property that fulfills when all input has been processed (to `true` if
+ * the source was exhausted or to `false` if iteration was aborted early),
+ * regardless of how many final iteration results have been consumed:
+ * ```
+ * await makeWorkPool(...).done;
+ * ```
+ */
+const makeWorkPool = <T, U = T, M extends 'all' | 'allSettled' = 'all'>(
+  source: AsyncIterable<T> | Iterable<T>,
+  config: undefined | [capacity: number][0] | { capacity?: number; mode?: M },
+  processInput: (
+    input: Awaited<T>,
+    index?: number,
+  ) => Promise<Awaited<U>> | Awaited<U> = x => x as any,
+): AsyncGenerator<
+  [
+    M extends 'allSettled' ? PromiseSettledResult<Awaited<U>> : Awaited<U>,
+    number,
+  ]
+> & { done: Promise<boolean> } => {
+  // Validate arguments.
+  if (isPrimitive(config)) config = { capacity: config, mode: undefined };
```

Comment:
```suggestion
  // Normalize config argument
  if (isPrimitive(config)) config = { capacity: config, mode: undefined };
  // Validate arguments.
```

but do we really need to support the primitive case? This code is pretty new; seems early to be maintaining backwards compatibility

Reply:
It's not for backwards compatibility, but for convenience and concision...
```js
              makeWorkPool($inputSource, $capacity, async input => {
                //…
              });
```
vs.
```js
              makeWorkPool(
                $inputSource,
                { capacity: $capacity },
                async input => {
                  //…
                },
              );
```
(noting the relative rarity of "allSettled")